### ISSUE TITLE
Negative Sampling

### ISSUE THREAD

**Issue by JuliaGast** [JuliaGast/TGB2#4]
- [x] Discuss what Negative Sampling approach to use
- [x] Integrate Negative Sampling (preliminary first version

**Comment by JuliaGast** [comment]
# Notes on negative sampling in the context of static KG link prediction.


## Negative sampling for computing the MRRs
Negative sampling for evaluation, i.e. for computing test MRRs is typically **NOT** done in static KG link prediction evaluation!

See info e.g. here:
![image](https://github.com/JuliaGast/TGB2/assets/41183430/7c884e05-0ffd-49ce-923e-2225697bc0ba)
[Start Small, Think Big: On Hyperparameter Optimization for Large-Scale Knowledge Graph Embeddings](https://www.uni-mannheim.de/media/Einrichtungen/dws/Files_People/Profs/rgemulla/publications/kochsiek22grash.pdf)
and here ([11] from above):
![image](https://github.com/JuliaGast/TGB2/assets/41183430/89065aae-06a6-4079-92ed-f2e3d9c35207)
[Parallel Training of Knowledge Graph Embedding Models: A Comparison of Techniques](https://www.vldb.org/pvldb/vol15/p633-kochsiek.pdf)

![image](https://github.com/JuliaGast/TGB2/assets/41183430/5b49d66c-7aba-4cae-8231-3de650723518)
[NBFNet](https://arxiv.org/pdf/2106.06935.pdf)

## Negative sampling during Training

Different commonly used approaches to train KGE models, which differ mainly in the way negative examples are generated

- Training with negative sampling (NegSamp) obtains for each positive triple t = (i, k, j) from the training data a set of (pseudo-)negative triples obtained by randomly perturbing the subject, relation, or object position in t (and optionally verifying that the so-obtained triples do not exist in the KG)
- 1vsAll: omit sampling and take all triples that can be obtained by perturbing the subject and object positions as negative examples for t (even if these tuples exist in the KG).  1vsAll is generally more expensive than NegSamp, but it is feasible (and even surprisingly fast in practice) if the number of entities is not excessively large. **(comment Julia: This is also done in TKG Forecasting Training)**
- KvsAll: (i) constructs batches from non-empty rows (i, k, ∗) or (∗, k, j) instead of from individual triples, and (ii) labels all such triples as either positive (occurs in training data) or negative (otherwise)
- [You can teach an old dog new tricks! (ICLR 2020)](https://openreview.net/pdf?id=BkxSmlBFvr)

For training with negative samples, different strategies exist:
- [Negative Sampling: Pykeen](https://pykeen.readthedocs.io/en/stable/reference/negative_sampling.html )(e.g.: Uniform, Bernoulli, Pseudotyped)
- [Negative Sampling in Knowledge Graph Representation Learning: A Review](https://www.researchgate.net/publication/378595890_Negative_Sampling_in_Knowledge_Graph_Representation_Learning_A_Review)

## Preliminary Conclusion (Up to discussion)
Due to the fact that it is not done in static KG completion, I opt to not use negative sampling for evaluation of TKG Forecasting, but instead compute the MRR based on scores for all entities in the KG.

In our datasets, we do not have more entities or significantly more test triples, and thus introducing negative sampling for evaluation of TKG models is not well motivated.


## Appendix: Datasets used for static KG link prediction
![image](https://github.com/JuliaGast/TGB2/assets/41183430/8ce5cbe3-315f-4058-9615-09bd26a25014)
(https://openreview.net/pdf?id=BkxSmlBFvr) 

![image](https://github.com/JuliaGast/TGB2/assets/41183430/57bf147a-d27b-4320-95e2-a9ed1ccd1fc9)
([ultra](https://arxiv.org/pdf/2310.04562.pdf))

## Appendix: Datasets used for TKG Forecasting 
![image](https://github.com/JuliaGast/TGB2/assets/41183430/589af8fa-e70e-4309-b787-92ae3d668211)

**Comment by shenyangHuang** [comment]
identify node types and ask the model only to predict MRR for nodes of the same type as the true answer? would that be a fair evaluation?

**Comment by JuliaGast** [comment]
not really; if models predict wrong node types (which happens) this would not be considered at all.
also, for many datasets node types are not given.

**Comment by JuliaGast** [comment]
- [ ] create negatives containing all nodes besides the nodes that are in same snapshot, i.e. n for n in (s,r,n,t) (time-aware filter setting).
solved with commit #12

**Comment by JuliaGast** [comment]
- [ ] create negative samples for inverse triples also 
related to #11

**Comment by JuliaGast** [comment]
suggestions of how to potentially choose node for negative samples:
combination of:
- entities that have been in a triple with node, relation in previous timesteps
- entities that have been in triple with this relation but not the same subject before (i.e.: same type)
- random entities

**Comment by JuliaGast** [comment]
What about negative sampling during training?

**Comment by JuliaGast** [comment]
Update:
After meeting with Michael on April 11th we decided to NOT do negative sampling, i.e. to do the 1-vs-all strategy.
